### ISSUE TITLE
High memory usage, slower speed with assembly

### ISSUE THREAD

**Issue by hickst** [clulab/reach#299]
I've been running the latest Reach master on the DR3 papers using BR 1.1.14 and Processors 5.9.4-SNAPSHOT.

Without assembly, all 15 papers finish in 5 minutes.
Unfortunately, running with assembly in 8G of RAM takes one hour (but works). The first 14 papers finish in 1/2 hour and the remaining paper takes an additional 1/2 hour. The last paper also produces the largest events and links files and the largest number of precedence features (but that's only 8 instances of precedence features).

When I increase the memory to 16G all papers finish in 25 minutes (with assembly).

I conclude that what we are seeing is that Assembly is slowing the processing of papers by a factor of 5x. Some input files, however, also cause a dramatic increase in memory usage.

**Comment by myedibleenso** [comment]
That's weird.  I can run them all on my laptop in ~15 minutes and that's with allocating less than 8G of ram.  

Anyway, I am aware that assembly is slowing things down.  As I mentioned several times, Tom, I profiled the system and the slowdown is from the feature-based classifier (the feature generation process is expensive).  We don't really have time to do a feature ablation before the evaluation to eliminate features without sacrificing performance and if we were to just disable the classifier altogether, we won't end up with any causal predecessors (our deterministic sieves are hardly ever triggered).  Perhaps I can lower the size of the ngram features.

**Comment by myedibleenso** [comment]
Try changing the 3 in these lines to a 2:

https://github.com/clulab/reach/blob/44ab46f4ae503eb466dcdeea7572b2baa0ed81d2/src/main/scala/org/clulab/assembly/relations/classifier/FeatureExtractor.scala#L61

https://github.com/clulab/reach/blob/44ab46f4ae503eb466dcdeea7572b2baa0ed81d2/src/main/scala/org/clulab/assembly/relations/classifier/FeatureExtractor.scala#L219

https://github.com/clulab/reach/blob/44ab46f4ae503eb466dcdeea7572b2baa0ed81d2/src/main/scala/org/clulab/assembly/relations/classifier/FeatureExtractor.scala#L222

**Comment by myedibleenso** [comment]
see the `leaner-features` branch.

**Comment by hickst** [comment]
> That's weird. I can run them all on my laptop in ~15 minutes and that's with allocating less than 8G of ram.

I'm surprised (and confused) by this since I've been very careful to thoroughly clean and recompile and keep variables constant, etc. I'm running on a 12-core system with 32G of RAM (of which 16G is available). We should make sure that we're using the same test conditions to report results:
*  Bioresources - released version 1.1.14
*  Processors - master branch 5.9.4-SNAPSHOT
*  Reach - master branch (1.3.2-SNAPSHOT)
*  Running all 15 DR3 files
*  Reach .sbtopts allocates  `-J-Xmx8G` (or less)
*  Unmodified, "out-of-the-box" application config (other than setting dir paths)
* Java 1.8.0_60-b27

**Comment by hickst** [comment]
The configuration in the `leaner-features` branch **is** helping to reduce the processing times:
```
621s vs. 1393s for the first 14 files to process
1126s vs. 2284s total time. (the slowest file is always PMC4729484.nxml)
```
The `leaner-features` branch is outputting the same number of precedence features and the other results seem very comparable, so a question here is: Can we live with the "leaner" features configuration?

In this sample of papers (the DR3 run) only one file out of 15 shows evidence of non-linear resource usage. So, is our DR3 sample representative of the paper population as a whole? If we have to process 100 (or 1K) papers, how many "resource hogs" can we expect to find in those populations and how much will they slow processing down?

**Comment by hickst** [comment]
One more comment: changing the `threadLimit` parameter in the application.conf file seems to affect *only* the paper reading phase. The actual processing of papers seems to be totally unaffected by this parameter. Is this expected behavior?

**Comment by myedibleenso** [comment]
>The leaner-features branch is outputting the same number of precedence features and the other results seem very comparable, so a question here is: Can we live with the "leaner" features configuration?

Yes, let's merge this.  We don't need trigrams.   Honestly, inclusion of trigrams was an accident and shouldn't have made it into `master`(my fault)...

>If we have to process 100 (or 1K) papers, how many "resource hogs" can we expect to find in those populations and how much will they slow processing down?

Dunno.  Why don't we try running ~100 random papers and see?

>only the paper reading phase
What do you mean by this?  I think parallelization throughout reach should be improved, but I believe that given the way it's currently written, the `processPaper` method should be getting called n (where `n` = `threadLimit`) times in parallel.

**Comment by hickst** [comment]
Well, I mean N threads start reading papers at once. After that, the verbose output is not very detailed but it sure looked like the papers finish processing at just the same intervals whether one uses 2 thread or 16 threads. I would not have expected that.

**Comment by myedibleenso** [comment]
>621s vs. 1393s for the first 14 files to process

How does this compare to Reach w/o assembly?

**Comment by hickst** [comment]
Well, I haven't done Reach w/o Assembly for the new `leaner-features` branch but the last run on DR3 with p5.9.4-SNAPSHOT w/o Assembly was 223s for the 14 files and 231s for all files (as in no significant difference for the evil PMC4729484 file to finish. In fact, it was not the last file processed in that run. PMC4729484 seems pretty clearly to be a troublemaker specifically for Assembly).
